programs/ido-program/programs/ido-program/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+use anchor_lang::{prelude::*, solana_program::clock::UnixTimestamp};
+use anchor_spl::token::{self, Burn, Mint, MintTo, Token, TokenAccount, Transfer};
+
+declare_id!("GYpxvUxtesyBSn69gnbfQChoUyJ7qdsG9nXS2Y2dQNH6");
+
+#[program]
+pub mod ido_program {
+    use super::*;
+
+    #[access_control(pre_ido_phase(start_ido_ts))]
+    pub fn initialize_pool(
+        ctx: Context<InitializePool>,
+        total_native_tokens: u64,
+        start_ido_ts: i64,
+        end_ido_ts: i64,
+        withdraw_deposit_token_ts: i64,
+        bump: u8,
+    ) -> Result<()> {
+        if !(start_ido_ts < end_ido_ts && end_ido_ts < withdraw_deposit_token_ts) {
+            return Err(ErrorCode::NonSequentialTimestamps.into());
+        }
+
+        if total_native_tokens == 0 {
+            return Err(ErrorCode::InvalidParameter.into());
+        }
+
+        let pool = &mut ctx.accounts.pool;
+
+        pool.pool_authority = *ctx.accounts.authority.key;
+        pool.redeemable_mint = ctx.accounts.redeemable_mint.key();
+        pool.native_mint = ctx.accounts.native_mint.key();
+        pool.deposit_token_mint = ctx.accounts.deposit_token_mint.key();
+        pool.pool_native = ctx.accounts.pool_native.key();
+        pool.pool_deposit_token = ctx.accounts.pool_deposit_token.key();
+        pool.total_native_tokens = total_native_tokens;
+        pool.start_ido_ts = start_ido_ts;
+        pool.end_ido_ts = end_ido_ts;
+        pool.withdraw_deposit_token_ts = withdraw_deposit_token_ts;
+
+        pool.bump = bump;
```

Comment:
Not a huge fan to let set the bump by an argument, it let place to bump seed non canonicalization exploit